File: day17/day17.py
from collections import namedtuple
from math import inf

Range = namedtuple("Range", ["min", "max"])

def read(name):
    def parse_range(s):
        s = s[2:]
        start, end = s.split("..")
        return Range(int(start), int(end))

    with open(name, 'r') as file:
        data = file.read().strip()
        data = data[len("target area: "):]
        xrange, yrange = map(parse_range, data.split(", "))
        return (xrange, yrange)

xtarget, ytarget = read('input.txt')

def sign(x):
    if x > 0:
        return 1
    if x < 0:
        return -1
    return 0

class Probe:
    def __init__(self, xtarget, ytarget, vx, vy):
        self.target = (xtarget, ytarget)
        self.pos = (0,0)
        self.v = (vx, vy)
        self.maxy = -inf
        self.hit = False
        self.overshot = False

    def step(self):
        self.pos = (self.pos[0]+self.v[0], self.pos[1]+self.v[1])
        self.v = (self.v[0]-sign(self.v[0]), self.v[1]-1)
        self.maxy = max(self.maxy, self.pos[1])

        x, y = self.target
        if self.pos[0] >= x.min and self.pos[0] <= x.max and self.pos[1] >= y.min and self.pos[1] <= y.max:
            self.hit = True
        if self.pos[0] > x.max or self.pos[1] < y.min:
            self.overshot = True

    def shoot(self):
        while (not self.hit) and (not self.overshot):
            self.step()
        if self.hit:
            return self.maxy
        return None

maxy = -inf
best = (0,0)
ngood = 0
for vx in range(1,1000):
    for vy in range(-1000,1000):
        my = Probe(xtarget, ytarget, vx, vy).shoot()
        if my is not None:
            ngood += 1
            if my > maxy:
                maxy = my
                best = (vx, vy)
print(f"Highest is {maxy} towards {best}")
print(f"There are {ngood} good vectors")

File: day17/day17.go
package main

import (
	"fmt"
	"math"
)

const (
	MinX = 257
	MaxX = 286
	MinY = -101
	MaxY = -57
)

func main() {
	var (
		numHits = 0
		highest = math.MinInt
	)
	for vx := 1; vx < 1000; vx++ {
		for vy := -1000; vy < 1000; vy++ {
			h, hit := Shoot(vx, vy)
			if hit {
				numHits++
			}
			if h > highest {
				highest = h
			}
		}
	}
	fmt.Printf("Highest Y is %d\n", highest)
	fmt.Printf("Number of hits is %d\n", numHits)
}

func Shoot(vx, vy int) (highest int, hit bool) {
	x, y := 0, 0
	highest = math.MinInt
	for x <= MaxX && y >= MinY {
		x += vx
		y += vy
		if vx > 0 {
			vx -= 1
		}
		vy -= 1

		if y > highest {
			highest = y
		}
		if x >= MinX && x <= MaxX && y >= MinY && y <= MaxY {
			return highest, true
		}
	}
	return math.MinInt, false
}
